File: backend/src/contour_controller.py
"""
Contour-following controller and height-field sampling.

This mirrors runtime-core's height_controller module for backend parity.
"""

from __future__ import annotations

from dataclasses import dataclass
import math


@dataclass
class HeightFieldParams:
    iterations: int = 64
    min_magnitude: float = 1.0e-6


@dataclass
class HeightFieldSample:
    height: float
    gradient: complex
    z: complex
    w: complex
    magnitude: float


@dataclass
class ContourControllerParams:
    correction_gain: float = 0.8
    projection_epsilon: float = 1.0e-6


@dataclass
class ContourStep:
    c: complex
    height: float
    height_error: float
    gradient: complex
    corrected_delta: complex


@dataclass
class ContourState:
    c: complex
    target_height: float
    last_delta: complex

    @classmethod
    def create(cls, c: complex, field_params: HeightFieldParams | None = None) -> "ContourState":
        params = field_params or HeightFieldParams()
        sample = sample_height_field(c, params)
        return cls(c=c, target_height=sample.height, last_delta=0j)

    def set_target_height(self, target_height: float) -> None:
        self.target_height = target_height

    def step(
        self,
        model_delta: complex,
        field_params: HeightFieldParams | None = None,
        controller_params: ContourControllerParams | None = None,
    ) -> ContourStep:
        params = field_params or HeightFieldParams()
        controller = controller_params or ContourControllerParams()
        sample = sample_height_field(self.c, params)
        height_error = sample.height - self.target_height
        corrected_delta = contour_correct_delta(
            model_delta,
            sample.gradient,
            height_error,
            controller,
        )
        self.c += corrected_delta
        self.last_delta = corrected_delta
        return ContourStep(
            c=self.c,
            height=sample.height,
            height_error=height_error,
            gradient=sample.gradient,
            corrected_delta=corrected_delta,
        )


def _dot(a: complex, b: complex) -> float:
    return a.real * b.real + a.imag * b.imag


def sample_height_field(c: complex, params: HeightFieldParams | None = None) -> HeightFieldSample:
    params = params or HeightFieldParams()
    z = 0j
    w = 0j
    for _ in range(params.iterations):
        w = 2.0 * z * w + 1.0
        z = z * z + c
    magnitude = abs(z)
    safe_mag = max(magnitude, params.min_magnitude)
    height = math.log(safe_mag)
    denom = max(safe_mag * safe_mag, params.min_magnitude * params.min_magnitude)
    conj_z = z.conjugate()
    g_complex = (conj_z * w) / denom
    gradient = complex(g_complex.real, -g_complex.imag)
    return HeightFieldSample(
        height=height,
        gradient=gradient,
        z=z,
        w=w,
        magnitude=magnitude,
    )


def contour_correct_delta(
    model_delta: complex,
    gradient: complex,
    height_error: float,
    params: ContourControllerParams | None = None,
) -> complex:
    params = params or ContourControllerParams()
    grad_norm_sq = gradient.real * gradient.real + gradient.imag * gradient.imag
    denom = grad_norm_sq + params.projection_epsilon
    normal_scale = _dot(gradient, model_delta) / denom
    delta_tangent = model_delta - gradient * normal_scale
    correction_scale = params.correction_gain * height_error / denom
    correction = gradient * correction_scale
    return delta_tangent - correction

File: frontend/src/lib/contourController.ts
import type { Complex } from './orbitSynthesizer';

export interface HeightFieldParams {
  iterations: number;
  minMagnitude: number;
}

export interface HeightFieldSample {
  height: number;
  gradient: Complex;
  z: Complex;
  w: Complex;
  magnitude: number;
}

export interface ContourControllerParams {
  correctionGain: number;
  projectionEpsilon: number;
}

export interface ContourState {
  c: Complex;
  targetHeight: number;
  lastDelta: Complex;
}

export interface ContourStep {
  c: Complex;
  height: number;
  heightError: number;
  gradient: Complex;
  correctedDelta: Complex;
}

const DEFAULT_HEIGHT_PARAMS: HeightFieldParams = {
  iterations: 64,
  minMagnitude: 1e-6
};

const DEFAULT_CONTOUR_PARAMS: ContourControllerParams = {
  correctionGain: 0.8,
  projectionEpsilon: 1e-6
};

function complexMul(a: Complex, b: Complex): Complex {
  return {
    real: a.real * b.real - a.imag * b.imag,
    imag: a.real * b.imag + a.imag * b.real
  };
}

function complexAdd(a: Complex, b: Complex): Complex {
  return { real: a.real + b.real, imag: a.imag + b.imag };
}

function complexScale(a: Complex, scale: number): Complex {
  return { real: a.real * scale, imag: a.imag * scale };
}

function complexSub(a: Complex, b: Complex): Complex {
  return { real: a.real - b.real, imag: a.imag - b.imag };
}

function complexMag(a: Complex): number {
  return Math.hypot(a.real, a.imag);
}

function complexConj(a: Complex): Complex {
  return { real: a.real, imag: -a.imag };
}

function dot(a: Complex, b: Complex): number {
  return a.real * b.real + a.imag * b.imag;
}

export function sampleHeightField(
  c: Complex,
  params: HeightFieldParams = DEFAULT_HEIGHT_PARAMS
): HeightFieldSample {
  let z: Complex = { real: 0, imag: 0 };
  let w: Complex = { real: 0, imag: 0 };

  for (let i = 0; i < params.iterations; i++) {
    w = complexAdd(complexScale(complexMul(z, w), 2.0), { real: 1, imag: 0 });
    z = complexAdd(complexMul(z, z), c);
  }

  const magnitude = complexMag(z);
  const safeMag = Math.max(magnitude, params.minMagnitude);
  const height = Math.log(safeMag);
  const denom = Math.max(safeMag * safeMag, params.minMagnitude * params.minMagnitude);
  const conjZ = complexConj(z);
  const gComplex = complexScale(complexMul(conjZ, w), 1.0 / denom);
  const gradient = { real: gComplex.real, imag: -gComplex.imag };

  return {
    height,
    gradient,
    z,
    w,
    magnitude
  };
}

export function contourCorrectDelta(
  modelDelta: Complex,
  gradient: Complex,
  heightError: number,
  params: ContourControllerParams = DEFAULT_CONTOUR_PARAMS
): Complex {
  const gradNormSq = gradient.real * gradient.real + gradient.imag * gradient.imag;
  const denom = gradNormSq + params.projectionEpsilon;
  const normalScale = dot(gradient, modelDelta) / denom;
  const deltaTangent = complexSub(modelDelta, complexScale(gradient, normalScale));
  const correctionScale = (params.correctionGain * heightError) / denom;
  const correction = complexScale(gradient, correctionScale);
  return complexSub(deltaTangent, correction);
}

export class ContourController {
  private state: ContourState;
  private fieldParams: HeightFieldParams;
  private controllerParams: ContourControllerParams;

  constructor(
    initialC: Complex,
    fieldParams: HeightFieldParams = DEFAULT_HEIGHT_PARAMS,
    controllerParams: ContourControllerParams = DEFAULT_CONTOUR_PARAMS
  ) {
    const sample = sampleHeightField(initialC, fieldParams);
    this.state = {
      c: initialC,
      targetHeight: sample.height,
      lastDelta: { real: 0, imag: 0 }
    };
    this.fieldParams = fieldParams;
    this.controllerParams = controllerParams;
  }

  setTargetHeight(targetHeight: number): void {
    this.state.targetHeight = targetHeight;
  }

  getTargetHeight(): number {
    return this.state.targetHeight;
  }

  getState(): ContourState {
    return { ...this.state };
  }

  step(modelDelta: Complex): ContourStep {
    const sample = sampleHeightField(this.state.c, this.fieldParams);
    const heightError = sample.height - this.state.targetHeight;
    const correctedDelta = contourCorrectDelta(
      modelDelta,
      sample.gradient,
      heightError,
      this.controllerParams
    );
    this.state.c = complexAdd(this.state.c, correctedDelta);
    this.state.lastDelta = correctedDelta;
    return {
      c: this.state.c,
      height: sample.height,
      heightError,
      gradient: sample.gradient,
      correctedDelta
    };
  }
}
